File: semanticdb-javac/src/main/java/com/sourcegraph/semanticdb_javac/ScipSymbols.java
package com.sourcegraph.semanticdb_javac;

import com.sourcegraph.semanticdb.SemanticdbSymbols;

/**
 * Converts SemanticDB-style symbol strings into the placeholder SCIP form expected by the
 * aggregator. The compiler plug-in doesn't know the final Maven coordinates, so globals are
 * prefixed with {@link #PLACEHOLDER_PREFIX} and the aggregator rewrites them into {@code "scip-java
 * maven <g> <a> <v> <descriptor>"}. Locals use the canonical {@code "local N"} form and pass
 * through unchanged.
 */
public final class ScipSymbols {

  public static final String PLACEHOLDER_PREFIX = ". . . . ";

  private ScipSymbols() {}

  public static String fromSemanticdbSymbol(String symbol) {
    if (symbol == null || symbol.isEmpty()) return "";
    if (SemanticdbSymbols.isLocal(symbol)) {
      return "local " + symbol.substring("local".length());
    }
    return PLACEHOLDER_PREFIX + symbol;
  }
}

File: semanticdb-kotlinc/src/main/kotlin/com/sourcegraph/semanticdb_kotlinc/ScipSymbols.kt
package com.sourcegraph.semanticdb_kotlinc

/**
 * Wraps descriptor strings into SCIP symbol strings: globals are prefixed with [PLACEHOLDER_PREFIX]
 * (rewritten to `scip-java maven <g> <a> <v> <descriptor>` once coordinates are known), locals use
 * the canonical `local N` form and pass through unchanged. Mirrors the Java `ScipSymbols` helper.
 */
object ScipSymbols {

    const val PLACEHOLDER_PREFIX: String = ". . . . "

    fun format(symbol: Symbol): String {
        if (symbol == Symbol.NONE) return ""
        val raw = symbol.toString()
        if (symbol.isLocal()) {
            return "local " + raw.substring("local".length)
        }
        return PLACEHOLDER_PREFIX + raw
    }
}
